File: backend/app/core/manutencao.py
"""Composicao dos textos do Relatorio de Manutencao.

Puro, sem I/O. Duas saidas alimentam o documento:
- "Tipo do Problema": os servicos escolhidos, em lista portuguesa;
- "Resumo do Servico": as frases padrao desses servicos, emendadas.

O Resumo do Servico e' GRAVADO na manutencao e fica congelado dali em diante —
e' prosa editada a mao, e editar o catalogo amanha nao pode reescrever um
resumo que alguem ja revisou.

Ja a linha do Tipo do Problema NAO e' gravada: ela deriva dos servicos ligados
a manutencao e e' recomposta a cada geracao do certificado, do mesmo jeito que
o certificado ja rederiva dados de cliente e aparelho toda vez que e' regerado.
"""


def _limpar(itens: list[str]) -> list[str]:
    return [i.strip() for i in itens if i and i.strip()]


def _sem_ponto_final(texto: str) -> str:
    return texto[:-1] if texto.endswith(".") else texto


def compor_problema(descricoes: list[str]) -> str:
    """Lista portuguesa dos servicos, terminada em ponto.

    Um -> "A."   Dois -> "A e B."   Tres ou mais -> "A, B e C."
    """
    itens = [_sem_ponto_final(d) for d in _limpar(descricoes)]
    if not itens:
        return ""
    if len(itens) == 1:
        corpo = itens[0]
    else:
        corpo = f"{', '.join(itens[:-1])} e {itens[-1]}"
    return f"{corpo}."


def _descrever_aparelho(modelo: str | None, serie: str | None) -> str:
    """Modelo e serie do aparelho, aguentando cadastro incompleto.

    Sem tratamento, um aparelho sem serie produziria "equipamento Mercury / nº de
    série ," — frase quebrada num documento que vai para o cliente.
    """
    partes = [p for p in (
        (modelo or "").strip(),
        f"nº de série {serie.strip()}" if (serie or "").strip() else "",
    ) if p]
    return " / ".join(partes) if partes else "não identificado"


def compor_resumo(modelo: str | None, serie: str | None,
                  servicos: list[tuple[str | None, str]]) -> str:
    """Texto padrao do "Resumo do Servico".

    O aparelho e a frase de conformidade aparecem UMA vez; so os servicos se
    repetem. Emendar uma frase completa por servico repetia os dois a cada item
    e, com tres ou mais, o resumo ficava longo e confuso.
    """
    itens = [(codigo, descricao.strip()) for codigo, descricao in servicos if descricao and descricao.strip()]
    if not itens:
        return ""
    lista = "; ".join(
        f"{codigo.strip()} – {descricao}" if (codigo or "").strip() else descricao
        for codigo, descricao in itens
    )
    rotulo = "referente ao serviço" if len(itens) == 1 else "referente aos serviços"
    return (
        f"Foi realizada a manutenção no equipamento {_descrever_aparelho(modelo, serie)}, "
        f"em conformidade com os procedimentos técnicos da Health & Safety, "
        f"{rotulo}: {lista}."
    )

File: frontend/src/app/ordens/manutencao.ts
import { apiJson, apiFetch, ApiError } from '../../lib/api'

// DELETE/204 não tem corpo — apiJson faz res.json() e quebraria. Mesmo padrão
// usado em propostas/api.ts, caixas/api.ts e cadastros/api.ts.
async function apiVoid(path: string, options: RequestInit = {}): Promise<void> {
  const res = await apiFetch(path, {
    ...options,
    headers: { 'Content-Type': 'application/json', ...(options.headers as Record<string, string>) },
  })
  if (!res.ok) {
    let detail = res.statusText
    try {
      const body = (await res.json()) as { detail?: string }
      if (body.detail) detail = body.detail
    } catch {
      // sem corpo JSON
    }
    throw new ApiError(res.status, detail)
  }
}

export interface ServicoManutencao {
  id: number
  /** SKU no catálogo comercial (`servicos`); nulo quando cadastrado à mão. */
  codigo: string | null
  descricao: string
  resumo_padrao: string
  ativo: boolean
}

export interface ManutencaoServicoItem {
  servico: number
  codigo: string | null
  descricao: string
  resumo_padrao: string
}

export interface Manutencao {
  id: number
  os: number
  numero: string | null
  data_manutencao: string | null
  resumo: string | null
  servicos: ManutencaoServicoItem[]
}

export interface ManutencaoPayload {
  numero: string | null
  data_manutencao: string | null
  resumo: string | null
  servicos: number[]
}

/** Modelo e série do aparelho, aguentando cadastro incompleto.
 *  Sem isso, aparelho sem série produziria "equipamento Mercury / nº de série ,". */
function descreverAparelho(modelo: string | null, serie: string | null): string {
  const partes = [
    (modelo ?? '').trim(),
    (serie ?? '').trim() ? `nº de série ${(serie ?? '').trim()}` : '',
  ].filter((p) => p !== '')
  return partes.length ? partes.join(' / ') : 'não identificado'
}

/** Espelha `compor_resumo` do backend (app/core/manutencao.py), para o modal
 *  mostrar o texto antes de salvar. Quem decide o valor final é a API.
 *
 *  O aparelho e a frase de conformidade aparecem UMA vez; só os serviços se
 *  repetem — emendar uma frase completa por serviço repetia os dois a cada item
 *  e ficava longo e confuso com três ou mais. */
export function comporResumo(
  modelo: string | null,
  serie: string | null,
  servicos: { codigo: string | null; descricao: string }[],
): string {
  const itens = servicos
    .map((s) => ({ codigo: (s.codigo ?? '').trim(), descricao: s.descricao.trim() }))
    .filter((s) => s.descricao !== '')
  if (itens.length === 0) return ''
  const lista = itens
    .map((s) => (s.codigo ? `${s.codigo} – ${s.descricao}` : s.descricao))
    .join('; ')
  const rotulo = itens.length === 1 ? 'referente ao serviço' : 'referente aos serviços'
  return `Foi realizada a manutenção no equipamento ${descreverAparelho(modelo, serie)}, `
    + `em conformidade com os procedimentos técnicos da Health & Safety, `
    + `${rotulo}: ${lista}.`
}

export const manutencaoApi = {
  obter: (osId: number): Promise<Manutencao> =>
    apiJson<Manutencao>(`/ordens/${osId}/manutencao`),
  salvar: (osId: number, payload: ManutencaoPayload): Promise<Manutencao> =>
    apiJson<Manutencao>(`/ordens/${osId}/manutencao`, { method: 'PUT', body: JSON.stringify(payload) }),
  listarServicos: (): Promise<ServicoManutencao[]> =>
    apiJson<ServicoManutencao[]>('/manutencao-servicos'),
  criarServico: (dados: { codigo?: string | null; descricao: string; resumo_padrao: string }): Promise<ServicoManutencao> =>
    apiJson<ServicoManutencao>('/manutencao-servicos', { method: 'POST', body: JSON.stringify(dados) }),
  atualizarServico: (id: number, dados: Partial<ServicoManutencao>): Promise<ServicoManutencao> =>
    apiJson<ServicoManutencao>(`/manutencao-servicos/${id}`, { method: 'PUT', body: JSON.stringify(dados) }),
  excluirServico: (id: number): Promise<void> =>
    apiVoid(`/manutencao-servicos/${id}`, { method: 'DELETE' }),
}
